File: src/main/scala/com/rephilo/luandun/leetcode/_202/Solution.scala
package com.rephilo.luandun.leetcode._202

import scala.annotation.tailrec

object Solution {
  /**
    * 202
    * 弗洛伊德循环查找算法
    * 生成一个链表，快指针走两格，慢指针走一格，如果汇合说明是循环，如果快指针先到1则是快乐数
    *
    * @param n
    * @return
    */
  def isHappy(n: Int): Boolean = {
    @tailrec
    def squareSum(n: Int, sum: Int): Int = {
      n match {
        case 0 => sum
        case _ => squareSum(n / 10, sum + (n % 10) * (n % 10))
      }
    }

    @tailrec
    def floydCycleFinding(slow: Int, fast: Int): Boolean = {
      fast match {
        case 1 => true
        case _ if fast == slow => false
        case _ => floydCycleFinding(squareSum(slow, 0), squareSum(squareSum(fast, 0), 0))
      }
    }

    floydCycleFinding(n, squareSum(n, 0))
  }
}

File: src/main/java/com/rephilo/luandun/leetcode/_202/Solution.java
package com.rephilo.luandun.leetcode._202;

public class Solution {
    /**
     * 202
     * 这个题有问题 为什么会收敛到一个循环，无法证明
     *
     * @param n
     * @return
     */
    public boolean isHappy(int n) {
        try {
            return duCalc(n, 1);
        } catch (Exception ignored) {

        }
        return false;
    }

    private Boolean duCalc(int m, int times) {
        if (times > 100) {
            return false;
        }
        int result = 0;
        while (m > 0) {
            int i = m % 10;
            result += i * i;
            m = m / 10;
        }

        if (result == 1) {
            return true;
        } else {
            return duCalc(result, times + 1);
        }
    }
}
